### DISCUSSION TITLE
Help in regards to porting to STM32

### DISCUSSION THREAD

**Discussion by patkepa** [eModbus/eModbus#331] [Q&A]
Hello!
I've got a question - from the experience of the devs, would it be plausible to attempt to port this ESP32 project for Arduino OPTA (STM32H747XI)? I'm searching for good async library for Modbus that I could use with my project, but sadly I cannot find anything, and this looks like the best implementation there is. If I could ask, are there any tips if I were to attempt to port this project for H7 Portenta architecture?

Thanks in advance!

**Comment by bertmelis** [comment]
Does the STM32 Arduino core run on FreeRTOS? If so, I suspect it is somewhat straightforward to port this library. Async will be a lot harder obviously because this is specific to ESP32/ESP8266.

Is async really needed? eModbus spawns a separate task to do its work so it doesn't interfere with your Arduino code.

**Reply by patkepa** [reply]
Thanks for quick answer, it seems that yes STM32 Arduino Core runs on FreeRTOS - in regards to the project I definitely require async, since the requirement for the project is simultaneous handling of 5 Modbus TCP clients in form of TCP-RTU gateway with additional web GUI. 

Do you know which parts of the library should I focus on in that case? Is it even doable without a major rewrite?

**Reply by Miq1** [reply]
Provided there is really RTOS running on it, it should work almost out-of-the-box with a ``ModbusBridgeEthernet`` with a ``ModbusClientRTU`` attached to it. Each client connection gets a dedicated task, so the only potential bottleneck is the RTU communication that is serialized.
